File: backend/dispute/src/main/java/com/commonsengine/dispute/service/DisputeService.java
package com.commonsengine.dispute.service;

import com.commonsengine.dispute.domain.Model.AiScreeningResult;
import com.commonsengine.dispute.domain.Model.ArbitrationVerdict;
import com.commonsengine.dispute.domain.Model.Dispute;
import com.commonsengine.dispute.domain.Model.DisputeId;
import com.commonsengine.dispute.domain.Model.DisputePriority;
import com.commonsengine.dispute.domain.Model.DisputeStatus;
import com.commonsengine.dispute.domain.Model.DisputeType;
import com.commonsengine.dispute.domain.Model.VerdictType;
import com.commonsengine.dispute.infrastructure.persistence.DisputeEntity;
import com.commonsengine.dispute.infrastructure.persistence.DisputeMapper;
import com.commonsengine.dispute.infrastructure.persistence.DisputeRepository;
import com.commonsengine.platform.exception.NotFoundException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;

/**
 * 纠纷工单业务服务（#65）
 *
 * 职责：
 * 1. 创建纠纷工单（FILED）并触发 AI 预审
 * 2. AI 预审（置信度阈值决定是否转人工）
 * 3. 人工仲裁（UNDER_REVIEW → ARBITRATING → RESOLVED）
 * 4. 查询工单列表
 *
 * 状态机校验由 Dispute.transitionTo 保证。
 */
@Service
public class DisputeService {

    private static final Logger log = LoggerFactory.getLogger(DisputeService.class);

    /** AI 预审置信度阈值——低于此值强制转人工。 */
    private static final double AI_CONFIDENCE_THRESHOLD = 0.7;

    private final DisputeRepository repository;
    private final DisputeMapper mapper;

    public DisputeService(DisputeRepository repository, DisputeMapper mapper) {
        this.repository = repository;
        this.mapper = mapper;
    }

    @Transactional
    public Dispute fileDispute(String transactionId, String consumerId, String workerId,
                                DisputeType type, DisputePriority priority, String description) {
        Dispute dispute = new Dispute(
                DisputeId.random(),
                transactionId,
                consumerId,
                workerId,
                type,
                priority,
                description,
                DisputeStatus.FILED,
                null,
                null,
                Instant.now(),
                null
        );
        DisputeEntity saved = repository.save(mapper.toEntity(dispute));
        log.info("纠纷工单已创建 id={} type={} priority={}", dispute.getId(), type, priority);
        return mapper.toDomain(saved);
    }

    /**
     * AI 预审——记录结果并决定是否转人工。
     *
     * @return 更新后的 Dispute（状态可能为 UNDER_REVIEW 或 RESOLVED）
     */
    @Transactional
    public Dispute recordAiScreening(DisputeId id, double confidence, String category,
                                      String summary) {
        Dispute dispute = loadOrThrow(id);
        dispute.transitionTo(DisputeStatus.AI_SCREENING);

        AiScreeningResult screening = new AiScreeningResult(confidence, category, summary,
                confidence < AI_CONFIDENCE_THRESHOLD);
        dispute.setAiScreening(screening);

        if (confidence >= AI_CONFIDENCE_THRESHOLD) {
            // 高置信度可直接转人工审核（后续可能自动解决）
            dispute.transitionTo(DisputeStatus.UNDER_REVIEW);
        } else {
            // 低置信度强制人工介入
            dispute.transitionTo(DisputeStatus.UNDER_REVIEW);
        }

        DisputeEntity saved = repository.save(mapper.toEntity(dispute));
        log.info("AI 预审完成 id={} confidence={} needsHuman={}", id, confidence,
                screening.needsHumanReview());
        return mapper.toDomain(saved);
    }

    /**
     * 提交仲裁裁决——状态必须为 UNDER_REVIEW 或 ARBITRATING。
     */
    @Transactional
    public Dispute arbitrate(DisputeId id, VerdictType verdictType, double amount, String reason) {
        Dispute dispute = loadOrThrow(id);

        if (dispute.getStatus() == DisputeStatus.UNDER_REVIEW) {
            dispute.transitionTo(DisputeStatus.ARBITRATING);
        }
        dispute.transitionTo(DisputeStatus.RESOLVED);

        ArbitrationVerdict verdict = new ArbitrationVerdict(verdictType, amount, reason, Instant.now());
        dispute.setVerdict(verdict);
        dispute.setResolvedAt(Instant.now());

        DisputeEntity saved = repository.save(mapper.toEntity(dispute));
        log.info("仲裁裁决完成 id={} verdict={} amount={}", id, verdictType, amount);
        return mapper.toDomain(saved);
    }

    @Transactional(readOnly = true)
    public Dispute findById(DisputeId id) {
        return loadOrThrow(id);
    }

    @Transactional(readOnly = true)
    public List<Dispute> findByStatus(DisputeStatus status) {
        return repository.findByStatusOrderByCreatedAtAsc(status).stream()
                .map(mapper::toDomain)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<Dispute> findByStakeholder(String consumerId, String workerId) {
        return repository.findByStakeholder(consumerId, workerId).stream()
                .map(mapper::toDomain)
                .toList();
    }

    private Dispute loadOrThrow(DisputeId id) {
        return repository.findById(id.value())
                .map(mapper::toDomain)
                .orElseThrow(() -> new NotFoundException("Dispute", id.value()));
    }
}

File: backend/dispute/src/main/kotlin/com/commonsengine/dispute/service/DisputeService.kt
package com.commonsengine.dispute.service

import com.commonsengine.dispute.domain.AiScreeningResult
import com.commonsengine.dispute.domain.ArbitrationVerdict
import com.commonsengine.dispute.domain.Dispute
import com.commonsengine.dispute.domain.DisputeId
import com.commonsengine.dispute.domain.DisputePriority
import com.commonsengine.dispute.domain.DisputeStatus
import com.commonsengine.dispute.domain.DisputeType
import com.commonsengine.dispute.domain.VerdictType
import com.commonsengine.dispute.infrastructure.persistence.DisputeEntity
import com.commonsengine.dispute.infrastructure.persistence.DisputeRepository
import com.commonsengine.dispute.infrastructure.persistence.toDomain
import com.commonsengine.dispute.infrastructure.persistence.toEntity
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant

/**
 * 纠纷仲裁服务
 *
 * 流程：提交 → AI 初筛 → (自动解决 | 调查 → 仲裁) → 解决
 *
 * 设计原则（架构文档 3.4 条）：
 * 1. AI 初筛 + 人工仲裁：AI 做初步分类和证据整理，复杂纠纷由仲裁委员会处理
 * 2. 透明流程：仲裁规则公开，劳动者有权申诉、有权知道仲裁结果的理由
 * 3. 仲裁委员会由劳动者代表、用户代表和社区代表组成
 *
 * 持久化：使用 JPA + PostgreSQL，重启不丢数据。
 */
@Service
open class DisputeService(
    private val repository: DisputeRepository,
) {

    /** 提交纠纷工单 */
    @Transactional
    open fun file(
        transactionId: String,
        filedBy: String,
        filedAgainst: String,
        type: DisputeType,
        description: String,
        evidenceUrls: List<String> = emptyList(),
    ): Dispute {
        val dispute = Dispute(
            id = DisputeId.random(),
            transactionId = transactionId,
            filedBy = filedBy,
            filedAgainst = filedAgainst,
            type = type,
            description = description,
            evidenceUrls = evidenceUrls,
        )
        repository.save(dispute.toEntity())
        return dispute
    }

    /**
     * AI 初筛——对纠纷进行分类、评级和证据摘要
     *
     * MVP 阶段使用规则匹配，后续接入 AI 服务层 NLP 模型。
     */
    @Transactional
    open fun aiScreening(disputeId: DisputeId): AiScreeningResult {
        val entity = findByDisputeIdOrThrow(disputeId)
        val dispute = entity.toDomain()

        updateEntityStatus(entity, DisputeStatus.AI_REVIEW)

        // 规则匹配：根据类型和描述关键词判断优先级
        val priority = when {
            dispute.type in listOf(DisputeType.BEHAVIORAL, DisputeType.DAMAGE_CLAIM) ->
                DisputePriority.HIGH
            dispute.type in listOf(DisputeType.FARE_DISPUTE, DisputeType.PAYMENT_ISSUE) ->
                DisputePriority.MEDIUM
            else -> DisputePriority.LOW
        }

        // 简单纠纷可以自动解决（低优先级 + 有明确证据）
        val canAutoResolve = priority == DisputePriority.LOW && dispute.evidenceUrls.isNotEmpty()

        // 证据摘要（MVP：截取描述前 200 字）
        val summary = if (dispute.description.length > 200) {
            dispute.description.take(200) + "..."
        } else {
            dispute.description
        }

        val result = AiScreeningResult(
            disputeId = disputeId,
            canAutoResolve = canAutoResolve,
            suggestedPriority = priority,
            summary = summary,
            category = dispute.type,
            confidence = if (canAutoResolve) 0.85 else 0.60,
            reasoning = buildString {
                append("AI 初筛：类型=${dispute.type}，")
                append("建议优先级=$priority。")
                if (canAutoResolve) {
                    append("低优先级且有证据，建议自动解决。")
                } else {
                    append("需人工调查，转入调查流程。")
                }
            },
        )

        // 根据初筛结果更新状态
        if (canAutoResolve) {
            updateEntityStatus(entity, DisputeStatus.AUTO_RESOLVED)
        } else {
            updateEntityStatus(entity, DisputeStatus.INVESTIGATION)
        }

        return result
    }

    /**
     * 仲裁——由仲裁委员会做出裁决
     *
     * 仲裁委员会由劳动者代表、用户代表和社区代表组成（比例见治理细则）。
     */
    @Transactional
    open fun arbitrate(
        disputeId: DisputeId,
        verdict: VerdictType,
        reasoning: String,
        compensationAmount: java.math.BigDecimal? = null,
    ): ArbitrationVerdict {
        val entity = findByDisputeIdOrThrow(disputeId)
        val dispute = entity.toDomain()

        require(dispute.status in listOf(DisputeStatus.INVESTIGATION, DisputeStatus.AI_REVIEW)) {
            "工单必须处于调查或初筛阶段才能仲裁，当前: ${dispute.status}"
        }

        updateEntityStatus(entity, DisputeStatus.ARBITRATION)

        val arbVerdict = ArbitrationVerdict(
            disputeId = disputeId,
            verdict = verdict,
            reasoning = reasoning,
            compensationAmount = compensationAmount,
        )

        // 直接修改已有 entity 并保存——保留数据库主键
        entity.status = DisputeStatus.RESOLVED
        entity.updatedAt = Instant.now()
        entity.resolution = "裁决：$verdict。理由：$reasoning"
        entity.resolvedAt = Instant.now()
        repository.save(entity)

        return arbVerdict
    }

    /** 查询工单 */
    @Transactional(readOnly = true)
    open fun findById(id: DisputeId): Dispute? =
        repository.findByDisputeId(id.value)?.toDomain()

    /** 查询全部工单 */
    @Transactional(readOnly = true)
    open fun findAll(): List<Dispute> =
        repository.findAll().map { it.toDomain() }

    /** 按状态查询 */
    @Transactional(readOnly = true)
    open fun findByStatus(status: DisputeStatus): List<Dispute> =
        repository.findByStatus(status).map { it.toDomain() }

    private fun findByDisputeIdOrThrow(disputeId: DisputeId): DisputeEntity =
        repository.findByDisputeId(disputeId.value)
            ?: throw IllegalArgumentException("纠纷工单不存在: $disputeId")

    private fun updateEntityStatus(entity: DisputeEntity, status: DisputeStatus) {
        entity.status = status
        entity.updatedAt = Instant.now()
        repository.save(entity)
    }
}
